security/mobile-app-network-security.md

```diff
@@ -0,0 +1,120 @@
+
+To prevent this type of attacks, a technique called "Certificate Pinning" is being used. The idea is that for most of the applications, they will talk to backend servers that are controlled by the same developers. Given the fact that each certificate is unique, at the application build time, the developers can generate "fingerprints" of the certificates of their backend services and "pin" them in the app. Then when the application is trying to establish the HTTPS connection, in addition to the default certificate checks, the HTTP client will also compute the "fingerprints" of the certificates and compare it againsted the "pinned" ones. If they are not matching, then the HTTP client knows that it is actually not talking to the real backend server, and thus refuse to setup the connection.
+
+## Solutions
+
+### Options
+
+Generally speaking, the above problems can be solved by 2 approaches:
+
+1. The first approach is to use the solution provided by the OS. 
+
+    For example, Android provides [Network Security Configurations](https://developer.android.com/training/articles/security-config.html) to allow developers to configure what CA store the apps can trust, and the "fingerprints" of their backend services.
+
+    The advantage of this approach is that there are no code changes required, and provides a center location for all the network security related configurations. Once configured, all the network libaries will respect the settings.
+
+    However, the biggest problem with this approach is the lack of standards accross different OSes. For example, iOS doesn't have the equivalent implementation for this feature. Even for Android itself, this feature is only supported on apps targeting Android 6.0 and above.
+
+2. The second approach is to use the APIs provided by the HTTP libaries. 
+
+    Most of the modern HTTP libaries allow developers to configure many different aspects of the HTTPS connections, and make it possible to support self-signed certificates and certificate pinning.
+
+    The downside of this approach is it requires developers to make code changes to their apps, and changes need to be made to every HTTP libary the apps are using. However, it provides a good foundation to form a solution that can work across different platforms.
+
+### Proposed Solution
+
+Based on the information above, and the fact that we want to find a solution that can work across different platforms for AeroGear services, we propose the following solution:
+
+1. In AeroGear SDKs, the HTTP libaries (there is only one for each platform) will be configured using the SDK config file (`mobile-services.json`). There will be a new section like this:
+
+    ```
+    {
+      ...,
+      "https": {
+        "selfSignedCerts": {
+          "enabled": true,
+          "keystores": [{
+            "path": "assets/dev.keystore"
```

Comment:
More questions than review stuff (I have a knowledge gap here so excuse any nonsense):

* The `assets/dev.keystore` file is also generated for the user?
* Will this name/path be configurable by the end user?
* I guess the format of this file will be different per platform? What will it look like with something like Cordova? Do you see anything getting more complex here?

Reply:
good questions.

> The assets/dev.keystore file is also generated for the user?

yes, it will be. But only for certain platforms (mainly android).

> Will this name/path be configurable by the end user?

I would think so.

> I guess the format of this file will be different per platform? What will it look like with something like Cordova? Do you see anything getting more complex here?

I am only thinking about ios and android here. I don't think we will be able to support self-signed certs for Cordova at all. Mainly because we will not be able to modify the configurations for the HTTP client in the webview. 

 I will also update the proposal to answer those questions.

Reply:
Ah yep that makes sense on the Cordova one, thanks